File: mkan_customization/mkan_customization/doc_events/purchase_order.py
import frappe
from frappe import _
from frappe.utils import flt


def after_insert(self, method):
    self.custom_bid_tabulation_check = 1
    self.db_set("custom_bid_tabulation_check", 1)

    if not self.supplier:
        return

    supplier = frappe.get_doc("Supplier", self.supplier)

    missing_fields = []

    if not supplier.get("custom_default_iban_no"):
        missing_fields.append("Default IBAN No.")

    if not supplier.get("custom_section_materials_services"):
        missing_fields.append("Section, Materials, Services")

    if not supplier.get("tax_id"):
        missing_fields.append("Tax ID")

    if not supplier.get("custom_cr_number"):
        missing_fields.append("CR Number")

    if not supplier.get("primary_address"):
        missing_fields.append("Address")

    if missing_fields:
        frappe.throw(
            _("Supplier {0} is missing required data: {1}").format(
                self.supplier_name, ", ".join(missing_fields)
            )
        )
    
@frappe.whitelist()
def service_status_map(source_name):
    from frappe.model.mapper import get_mapped_doc

    # Step 1: Check if a Service Status already exists with this Purchase Order
    existing = frappe.get_all("Service Status", filters={"purchase_order": source_name}, fields=["name"])
    if existing:
        frappe.throw(f"Service Status <a href='/app/service-status/{existing[0].name}' style='color:blue;'>{existing[0].name}</a> already exists for this Purchase Order.")

    # Step 2: Map Purchase Order → Service Status
    def postprocess(source_doc, target_doc):
        target_doc.purchase_order = source_doc.name
        target_doc.customer = source_doc.customer

    return get_mapped_doc(
        "Purchase Order",
        source_name,
        {
            "Purchase Order": {
                "doctype": "Service Status",
                "validation": {
                    "docstatus": ["=", 1]
                }
            },
            "Purchase Order Item": {
                "doctype": "Service Status Details",
                "field_map": {
                    "item_code": "item",
                    "item_name": "item_name",
                    "description": "description",
                    "material_request": "material_request"
                },
                "add_if_empty": True
            }
        },
        postprocess=postprocess
    )


def validate_rates_against_bid_tabulation(doc):
    """
    Returns mismatch data — does NOT throw.
    Client side handles the dialog.
    """
    if not doc.get("bid_tabulation"):
        return []

    sq_name = frappe.db.get_value(
        "Bid Tabulation Discussion",
        doc.bid_tabulation,
        "supplier_quotation"
    )

    if not sq_name:
        return []

    sq_items = frappe.get_all(
        "Supplier Quotation Item",
        filters={"parent": sq_name},
        fields=["item_code", "rate"],
    )

    sq_rate_map = {row["item_code"]: flt(row["rate"]) for row in sq_items}

    mismatches = []

    for item in doc.items:
        sq_rate = sq_rate_map.get(item.item_code)

        if sq_rate is None:
            mismatches.append({
                "item_code": item.item_code,
                "item_name": item.item_name,
                "sq_rate": "Not in SQ",
                "po_rate": flt(item.rate),
            })
        elif flt(sq_rate) != flt(item.rate):
            mismatches.append({
                "item_code": item.item_code,
                "item_name": item.item_name,
                "sq_rate": sq_rate,
                "po_rate": flt(item.rate),
            })

    return mismatches


@frappe.whitelist()
def check_rate_mismatch(doc):

    doc = frappe.parse_json(doc)

    if not doc.get("bid_tabulation"):
        return []

    sq_name = frappe.db.get_value(
        "Bid Tabulation Discussion",
        doc.get("bid_tabulation"),
        "supplier_quotation"
    )

    if not sq_name:
        return []

    sq_items = frappe.get_all(
        "Supplier Quotation Item",
        filters={"parent": sq_name},
        fields=["item_code", "rate"],
    )

    sq_rate_map = {
        row["item_code"]: flt(row["rate"])
        for row in sq_items
    }

    mismatches = []

    for item in doc.get("items", []):

        sq_rate = sq_rate_map.get(item.get("item_code"))

        if sq_rate is None:
            continue

        if flt(sq_rate) != flt(item.get("rate")):

            mismatches.append({
                "item_code": item.get("item_code"),
                "item_name": item.get("item_name"),
                "sq_rate": sq_rate,
                "po_rate": flt(item.get("rate")),
            })

    return mismatches

File: mkan_customization/public/js/purchase_order.js
frappe.ui.form.on("Purchase Order", {
    refresh: function (frm) {
        render_rate_mismatch_warning(frm);
        frm.add_custom_button(
            __("Service Status"),
            () => {
                frappe.call({
                    method: "mkan_customization.mkan_customization.doc_events.purchase_order.service_status_map",
                    args: {
                        source_name: frm.doc.name
                    },
                    callback: function (response) {
                        if (response.message) {
                            frappe.model.sync(response.message);
                            frappe.set_route("Form", response.message.doctype, response.message.name);
                        }
                    }
                });
            },
            __("Create")
        );
        if (frm.doc.custom_bid_tabulation_check == 0) {
            frappe.call({
                method: "frappe.client.get_list",
                args: {
                    doctype: "Purchase Order",
                    filters: {
                        docstatus: ["<", 2],
                        bid_tabulation: frm.doc.bid_tabulation
                    },
                    fields: ["name"]
                },
                callback: function (response) {
                    let existing_docs = response.message;
                    if (existing_docs.length > 0) {
                        let po_links = existing_docs
                            .map(row => `<a href="/app/purchase-order/${row.name}" style="color:blue;">${row.name}</a>`)
                            .join("<br>");

                        frappe.confirm(
                            `The following Purchase Orders already exist for this Bid Tabulation:<br><br>${po_links}<br><br>Do you want to proceed?`,
                            () => {
                                frappe.validated = true;
                                frm.save();
                            },
                            () => {
                                frappe.validated = false;
                                if (frm.doc.bid_tabulation) {
                                    frappe.set_route("Form", "Bid Tabulation Discussion", frm.doc.bid_tabulation);
                                }
                            }
                        );

                    }
                }
            });
        }
    },

    validate(frm) {
        render_rate_mismatch_warning(frm);
    },

    items_on_form_rendered(frm) {
        render_rate_mismatch_warning(frm);
    }

});

async function render_rate_mismatch_warning(frm) {

    // HTML field must exist
    if (!frm.fields_dict.rate_mismatch_html) {
        return;
    }

    // No Bid Tabulation
    if (!frm.doc.bid_tabulation) {

        frm.fields_dict.rate_mismatch_html.$wrapper.html("");

        return;
    }

    // Fetch mismatches
    let r = await frappe.call({
        method: "mkan_customization.mkan_customization.doc_events.purchase_order.check_rate_mismatch",
        args: {
            doc: frm.doc
        }
    });

    const mismatches = r.message || [];

    // No mismatches
    if (!mismatches.length) {

        frm.fields_dict.rate_mismatch_html.$wrapper.html("");

        return;
    }

    // Build rows
    let rows = "";

    mismatches.forEach((m, idx) => {

        rows += `
            <tr>
                <td style="padding:10px;">
                    ${idx + 1}
                </td>

                <td style="padding:10px;">
                    <b>${m.item_code || ""}</b>
                </td>

                <td style="padding:10px;">
                    ${m.item_name || ""}
                </td>

                <td style="
                    padding:10px;
                    text-align:right;
                    color:#c0392b;
                    font-weight:600;
                ">
                    ${format_currency(m.sq_rate, frm.doc.currency)}
                </td>

                <td style="
                    padding:10px;
                    text-align:right;
                    color:#2471a3;
                    font-weight:600;
                ">
                    ${format_currency(m.po_rate, frm.doc.currency)}
                </td>
            </tr>
        `;
    });

    // Render HTML
    frm.fields_dict.rate_mismatch_html.$wrapper.html(`

        <div style="
            border:1px solid #f1c40f;
            border-radius:10px;
            background:#fffdf5;
            padding:16px;
            margin-top:10px;
            margin-bottom:15px;
        ">

            <div style="
                display:flex;
                align-items:center;
                gap:10px;
                margin-bottom:12px;
            ">

                <span style="
                    font-size:18px;
                ">
                    ⚠️
                </span>

                <div>

                    <div style="
                        font-size:15px;
                        font-weight:700;
                        color:#8a6d3b;
                    ">
                        Supplier Quotation Rate Mismatch
                    </div>

                    <div style="
                        font-size:13px;
                        color:#666;
                        margin-top:2px;
                    ">
                        Some Purchase Order item rates differ from the
                        linked Supplier Quotation rates.
                    </div>

                </div>

            </div>

            <div style="
                max-height:350px;
                overflow:auto;
                border:1px solid #eee;
                border-radius:6px;
            ">

                <table class="table table-bordered"
                    style="
                        margin:0;
                        width:100%;
                        font-size:13px;
                    ">

                    <thead>

                        <tr style="
                            background:#f8f9fa;
                        ">

                            <th style="padding:10px;">#</th>

                            <th style="padding:10px;">
                                Item Code
                            </th>

                            <th style="padding:10px;">
                                Item Name
                            </th>

                            <th style="
                                padding:10px;
                                text-align:right;
                            ">
                                SQ Rate
                            </th>

                            <th style="
                                padding:10px;
                                text-align:right;
                            ">
                                PO Rate
                            </th>

                        </tr>

                    </thead>

                    <tbody>
                        ${rows}
                    </tbody>

                </table>

            </div>

        </div>
    `);
}
